### PR TITLE
Prefer imports starting with std

### PR DESCRIPTION
**Pull request by SomeoneToIgnore**
Closes https://github.com/rust-analyzer/rust-analyzer/issues/2915

### CODE REVIEW

**Comment by SomeoneToIgnore** [comment]
I'd like to write a test on it, but could not reproduce the `alloc::sync::Arc`/`alloc::sync::Arc` structure in the test: if I try to use nested modules as `/std/sync/mod.rs crate:std deps:alloc`, I get panics despite the paths being correct (as if the crate name is not considered in the imports).
If I try to emulate the `pub mod sync` instead, I always get the `best_path == None` and cannot get into the right match branch in the method I've added.

Apparently, I'm doing something wrong, but I have no clue on how to make it proper. 
Any ideas?

**Comment by matklad** [comment]
@SomeoneToIgnore could you post an exact test  that does not work?

File: crates/ra_hir_def/src/find_path.rs [hunk 1]
```diff
@@ -111,15 +111,40 @@ fn find_path_inner(
...
+
+        if prefer_new_path(best_path_len, best_path.as_ref(), &new_path) {
+            best_path_len = path_len(&new_path);
+            best_path = Some(new_path);
         }
     }
     best_path
 }
 
+fn prefer_new_path(old_path_len: usize, old_path: Option<&ModPath>, new_path: &ModPath) -> bool {
```
Comment (matklad) [inline crates/ra_hir_def/src/find_path.rs]: Can we express this as a key function? like `(path.starts_with_std(), path.len())`?

**Comment by SomeoneToIgnore** [comment]
> @SomeoneToIgnore could you post an exact test that does not work?

I think I should stop writing the code late in the night (I wish I have any other time to do so for RA though... )
I've started writing an extended explanation to you and realised that all my assumptions about those tests are wrong.

For some reason, I've thought that lines `//- /sync/mod.rs crate:std deps:alloc` do something, as if they were forcing the files with the code below to be created or something.
Now I realize those are just comments.

I'll try to dig it more, thank you for reaching out to help me here.

**Comment by matklad** [comment]
>For some reason, I've thought that lines //- /sync/mod.rs crate:std deps:alloc do something

They very much do something: https://github.com/rust-analyzer/rust-analyzer/blob/0d6e5a986cc7c3a434c2107edf748d1c26ac3f69/crates/ra_db/src/fixture.rs#L152-L163

**Comment by SomeoneToIgnore** [comment]
Then I'm fully confused by their behavior (I know I should look in the test code better, but did not have the possibility to do that properly yet).

I try to replicate the Arc location in both crates and import it.

This test panics:

```
#[test]
fn why_does_it_panic() {
    let code = r#"
    //- /main.rs crate:main deps:std
    <|>
    //- /sync/mod.rs crate:std deps:alloc
    pub use alloc::sync::Arc;
    //- /sync.rs crate:alloc
    pub struct Arc;
    "#;
    check_found_path(code, "std::sync::Arc");
}
```

its simplified version also:

```
#[test]
fn why_does_it_panic() {
    let code = r#"
    //- /main.rs crate:main deps:std
    <|>
    //- /sync/mod.rs crate:std deps:alloc
    pub struct Arc;
    "#;
    check_found_path(code, "std::sync::Arc");
}
```

When I try to use `pub mod` instead, it "works:"

```
#[test]
fn zzzzz() {
    let code = r#"
    //- /main.rs crate:main deps:std
    <|>
    //- /sync/mod.rs crate:std deps:alloc
    pub mod sync {
        pub use alloc::sync::Arc;
    }
    //- /sync.rs crate:alloc
    pub mod sync {
        pub struct Arc;
    }
    "#;
    check_found_path(code, "std::sync::Arc");
}
```

but that is odd, since I basically duplicate module names in the file and in the `mod` declaration, from my experience with the Rust module system, the resulting import should be `std::sync::sync::Arc`.

And, anyway, when I put the debug prints into the `prefer_new_path` method, it never gets to any of the first two match arms, which is exactly what I want to test.
Moreover, `best_path` is always `None` for that test.

File: crates/ra_hir_def/src/find_path.rs (hunk 1 above)
Comment (SomeoneToIgnore) [reply crates/ra_hir_def/src/find_path.rs]: Yes, way more readable now, thanks.

**Comment by matklad** [comment]
Ah, well, it works exactly as expected! (but it took me a long time to realize this).

You want something like this:

```rust
        let code = r#"
        //- /main.rs crate:main deps:std
        <|>

        //- /std.rs crate:std deps:alloc
        pub use alloc::sync

        //- /alloc.rs crate:alloc
        pub mod sync {
            pub struct Arc;
        }
        "#;
```

files with `//-` are crate roots, you need to make sure they contain the module structure you want.

**Comment by SomeoneToIgnore** [comment]
Ok, so that's basically my 2nd option that does not go into the `if` branch I want :D
Thanks for the explanation anyway, I will keep searching for some test case that covers it.

**Comment by matklad** [comment]
Yeah, to clarify, if you write `crate:foo`, the file name is mostly irrelevant, as it becomes the root module of the crate.

The file name matters for finding child modules, but that's it

**Comment by flodiebold** [comment]
Also note that if your main crate in the test only has the dependency to `std`, the test doesn't actually reproduce the problem, since the main crate won't be able to refer to `alloc` at all.

**Comment by SomeoneToIgnore** [comment]
Well, simply adding it to the std dependencies did not help to enter the required branch, but thanks a lot, that's what I've been missing from the very start when tried to make it work in the night.

File: crates/ra_hir_def/src/find_path.rs [hunk 2]
```diff
@@ -112,23 +141,27 @@ fn find_path_inner(
...
-            PathKind::Crate => 1,
-            PathKind::Abs => 0,
-            PathKind::DollarCrate(_) => 1,
-        }
+fn select_best_path(old_path: ModPath, new_path: ModPath) -> ModPath {
+    if old_path.starts_with_std() && new_path.should_start_with_std() {
+        tested_by!(prefer_std_paths);
+        old_path
+    } else if new_path.starts_with_std() && old_path.should_start_with_std() {
+        tested_by!(prefer_std_paths);
```
Comment (matklad) [inline crates/ra_hir_def/src/find_path.rs]: :+1:

**Comment by SomeoneToIgnore** [comment]
And now it's ready to be reviewed, thank you again for the help.

**Comment by SomeoneToIgnore** [comment]
> bors r+

Judging by the look of the checks, bors decided to ignore this one :)

**Comment by matklad** [comment]
bors ping

**Comment by matklad** [comment]
bors merge

**Comment by flodiebold** [comment]
Bors doesn't seem to take commands from review comments anymore.

**Pull request merged**